[#42] 특정 사용자에게 메시지 보내기 기능 구현

msugo1 opened the pull request:
## 📖 Config 관련

### 📑  WebSocketConfig

- 사용자들이 다른 사용자에게 보낸 매시지가 실시간으로 반영되도록 WebSocket 설정 추가

## 📖 Controller 계층 관련

### 📑 ChatController

- 메시지 보내기 요청을 매핑해서 비즈니스 계층으로 넘겨 줄 `sendChat` 메소드 추가

- `setApplicationDestinationPrefixes`와 `messageMapping`의 URL을 조합해서 `/app/chat`으로 들어오는 요청을 처리

## 📖 Service 계층 관련

### 📑 ChatService

- 요청으로 들어온 메시지를 저장하고 (insertChat) 해당 사실을 메시지를 받는 사용자에게 알려주는 로직을 처리하는 `sendChat`  메소드 작성

- 컨트롤러 계층에서 chatBoxId를 `ChatSendRequest`를 매핑한 객체를 통해 전달 받음 -> chatBoxId가 존재하는지 확인

1. chatBoxId가 존재하는 경우

- 저장할 데이터를 한 곳에 모은 `ChatData` 객체를 생성한 뒤 데이터를 DB에  삽입하고 해당 사용자에게 알림을 전송

2. chatBoxId가 존재하지 않는 경우

- 두 사용자 간에 이용하는 chatBox의 Id를 DB에서 가져온다.
  -> 가져오기 전 기존의 ChatBox가 있는지 확인하고 있을 경우 새로 만들지 않고, 해당 ChatBoxId를 가져온다.
  -> 기존의 ChatBox가 존재하지 않고, 검증과증을 다 통과했으면 새로운 ChatBox를 만들어 ChatBoxId를 가져온다.

- 해당 로직을 처리할 createIfNotExist 메소드 및 fetchChatBoxId 메소드 추가

- `SimpMessagingTemplate`을 주입 받아 사용하며, 해당 클래스가 가지고 있는 `convertAndSendToUser` 메소드를 사용하여
  `/user/{recipient}/queue/message`의 형태로 알림을 전달할 수 있도록 매개변수 전달

## 📖 Mapper 계층 관련

### 📑 ChatMapper

- 사용자가 보낸 메시지를 DB에 저장 할 insertChat 메소드 작성

- 서비스 단의 로직을 DB와 연계하여 처리할 getChatBoxId, insertChatBox, isChatBoxExist 메소드 추가

- 해당 메소드에 대한 각각의 쿼리 추가

### 📑 ChatMapper.xml

- 해당 메소드를 처리할 쿼리 작성

## 📖 Model 관련

### 📑 ChatSendRequest

- 메시지를 보낼 사용자와 내용을 매핑할 객체를 생성할 클래스

- ChatBoxId도 함께 전달 받을 수 있도록 멤버 변수 추가

### 📑 ChatNotification

- 메시지가 왔다는 요청을 다시 전달하는데 사용할 객체를 생성할 클래스

### 📑 ChatData

- 필요한 데이터들을 멤버 변수로 하는 객체를 생성해서 DB에 전달해주는 DAO의 용도로 사용할 객체를 생성하는 클래스 추가

## 📖 SQL 파일 관련

- chats, chatboxes 관련 schema 작성 및 추가

## 📖 테스트 관련

### 📑 ChatServiceTest 

- ChatService 로직을 테스트 하는 단위 테스트 작성

Commit:
[#42] 특정 사용자에게 메시지 보내기 기능 구현

Config 관련

* WebSocketConfig

- 사용자들이 다른 사용자에게 보낸 매시지가 실시간으로 반영되도록
  WebSocket 설정 추가

Controller 계층 관련

* ChatController

- 메시지 보내기 요청을 매핑해서 비즈니스 계층으로 넘겨 줄 `sendChat` 메소드 추가

- `setApplicationDestinationPrefixes`와 `messageMapping`의 URL을
  조합해서 `/app/chat`으로 들어오는 요청을 처리

Service 계층 관련

* ChatService

- 요청으로 들어온 메시지를 저장하고 (insertChat) 해당 사실을 메시지를
  받는 사용자에게 알려주는 (notifyChat) 로직을 처리하는 `sendChat`
  메소드 작성

- `SimpMessagingTemplate`을 주입 받아 사용하며, 해당 클래스가 가지고
  있는 `convertAndSendToUser` 메소드를 사용하여
  `/user/{recipient}/queue/message`의 형태로 알림을 전달할 수 있도록
  notifyChat 메소드 작성

Mapper 계층 관련

* ChatMapper

- 사용자가 보낸 메시지를 DB에 저장 할 insertChat 메소드 작성

* ChatMapper.xml

- 해당 메소드를 처리할 쿼리 작성

Model 관련

* ChatSendRequest

- 메시지를 보낼 사용자와 내용을 매핑할 객체를 생성할 클래스

* ChatNotification

- 메시지가 왔다는 요청을 다시 전달하는데 사용할 객체를 생성할 클래스

Commit:
[#42] 메시지 전송 설계/로직 수정

* 개요
- DB에 메시지 함 관련 Schema를 추가해서 차후 DB로부터 메시지를 가져올
  때 쿼리가 효율적으로 동작할 수 있는 기반을 마련

Service 계층 관련

* ChatService

- sendChat 로직 수정

: 컨트롤러 계층에서 chatBoxId를 `ChatSendRequest`를 매핑한 객체를 통해
전달 받음 -> chatBoxId가 존재하는지 확인

1. chatBoxId가 존재하는 경우

- 저장할 데이터를 한 곳에 모은 `ChatData` 객체를 생성한 뒤 데이터를 DB에
  삽입하고 해당 사용자에게 알림을 전송

2. chatBoxId가 존재하지 않는 경우

- 두 사용자 간에 이용하는 chatBox의 Id를 DB에서 가져온다.
  -> 가져오기 전 기존의 ChatBox가 있는지 확인하고 있을 경우 새로 만들지 않고, 해당
  ChatBoxId를 가져온다.
  -> 기존의 ChatBox가 존재하지 않고, 검증과증을 다 통과했으면 새로운
  ChatBox를 만들어 ChatBoxId를 가져온다.

- 해당 로직을 처리할 createIfNotExist 메소드 및 fetchChatBoxId 메소드 추가

Mapper 계층 관련

* ChatMapper, ChatMapper.xml

- 서비스 단의 로직을 DB와 연계하여 처리할 getChatBoxId, insertChatBox, isChatBoxExist 메소드 추가

- 해당 메소드에 대한 각각의 쿼리 추가

Model 계층 관련

* ChatRequest

- ChatBoxId도 함께 전달 받을 수 있도록 멤버 변수 추가

* ChatData

- 필요한 데이터들을 멤버 변수로 하는 객체를 생성해서 DB에 전달해주는
  DAO의 용도로 사용할 객체를 생성하는 클래스 추가

SQL 관련

- chats, chatboxes 관련 schema 작성 및 추가

테스트 관련

- ChatService 로직을 테스트 하는 단위 테스트 작성

msugo1 commented:
Can one of the admins verify this patch?

Commit:
feature/41 오류 수정내역 반영

Commit:
불필요한 import 제거

f-lab-dev commented:
이 웹소켓 채팅 아키텍쳐는 draw.io로 그려서 위키에 업데이트 해보시면 좋을 것 같네요~
웹소켓을 사용한 이유와 함께요. 웹소켓 말고 대체할만한 방법이 있었는지, 그 방법에 비해 웹소켓이 어떤게 유리했던건지

f-lab-dev commented on src/main/java/me/soo/helloworld/service/ChatService.java:
```diff
@@ -0,0 +1,51 @@
+        }
+
+        ChatData chatData = ChatData.create(sender, chatBoxId, chatRequest.getRecipient(), chatRequest.getContent());
+        chatMapper.insertChat(chatData);
```
실시간 성능이 중요한 채팅을 보낼때마다 Blocking IO가 발생하는건 따로 한번 해결해주시는게 좋을 것 같네요~

msugo1 replied:
카프카 공부하고 도입해서 새로운 이슈로 PR 보냈습니다!

Commit:
[#43] 메시지 실시간 처리향상을 위한 카프카 도입

* pom.xml

- 카프카 의존성 추가

Config 관련

* KafaConfigProducer

- 카프카 프로듀서에 대한 설정 추가

1. acks를 1로 설정하여 메시지 발송에 대한 안정성을 가지도록 설정

2. batch size를 적절히 증가시키고, linger ms 타임을 부여하여 메시지의
   단일 전송으로 인한 네트워크 비용을 줄일 수 있도록 설정

* KafkaConfigConsumer

- 카프카 컨슈머에 대한 설정 추가

1. auto commit을 false로 설정하여 배치로 메시지를 가져와서 처리하는데
   실패하면 다시 커밋되지 않은 offset 위치로 돌아갈 수 있게 설정

2. auto offset reset을 earliest로 설정하여 offset이 없는 경우 메시지가
   처음부터 처리될 수 있도록 설정

3. min fetch size와 fetch max wait 타임을 설정하여 일정한 시간이
   지나거나 일정 양의 데이터를 확보하는 경우에만 메시지를 가져오도록
   하여 네트워크 비용을 줄이도록 설정

4. batch listener 값을 true로 변경하여 한 번에 많은 양의 데이터를 가져올
   수 있도록 설정하고, 오류가 발생했을 때 RecoveringBatchErrorHandler를
   활용해서 다시 시도할 수 있도록 설정

Service 계층 관련

* ChatService

- 개별 메시지 요청이 들어올 때마다 DB에 바로 저장하여 블로킹 IO를
  유발했던 것을 카프카로 먼저 묶음으로 보낸 뒤 다시 카프카 컨슈머가
  묶음으로 가져와 처리할 수 있는 구조로 변경하여 DB로의 접근과 블로킹 IO를 최대한 줄이도록 노력

* KafkaProducerService

- 프로듀서 Config 설정하면서 생성한 카프카 템플릿을 주입받아 카프카로
  메시지를 보내는 send 메소드 추가

- 콜백을 추가하여 비동기 전송 방식으로 메시지를 보내고 해당 결과를 받아 처리할 수 있도록 설정

* KafkaConsumerService

- 배치형식으로 메시지를 카프카에서 받아와서 데이터베이스에 bulk insert를
  통해 메시지를 다량으로 저장하는 비즈니스 로직을 처리하는 saveChats
  메소드 설정

- @KafkaListener 어노테이션을 추가해 메시지를 저장한 토픽으로부터
  메시지를 받아올 수 있도록 설정

Model 관련

* ChatSendRequest

- 메시지가 서버로 들어오면 바로 저장하는 것이 아니라 카프카를 거쳐
  차후에 DB에 저장하는 방식이므로 시간차를 고려해 클라이언트로부터
  처음에 메시지를 보낸시간을 받아오도록 변경

* ChatWrite

- 메시지를 보낸 시간을 추가하고 카프카/DB로 전송할 모든 내용을 담은
  인스턴스를 생성하기 위한 클래스 추가

Mapper 계층 관련

* ChatMapper/ChatMapper.xml

- 카프카에서 가져온 메시지 목록을 한번에 bulk insert하기 위한
  insertChats 메소드 추가와 쿼리 추가

테스트 관련

- 로직 수정으로 인한 테스트 코드 수정

- 포스트맨 및 미리 작성해둔 클라이언트 코드를 통해 메시지가 상대방에게
  잘 전달되고, 카프카로 전달한 메시지가 다시 DB에 잘 저장되는지 확인

Commit:
[#43] 피드백 반영

- 불필요하게 많이 생성될 수 있는 로그 라인 제거

msugo1 committed:
Merge pull request #92 from f-lab-edu/feature/43

[#43] 메시지 실시간 처리성능 향상을 위한 카프카 도입
